# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

Создаю функцию, в которой при подстановке столбца, по которому будет высчитываться количество кредитополучателей, количество должников и доля должников для ответа на следующие вопросы. Нужно будет ввести датафрейм, название столбца, по которому группируются данные, и название индекса.

In [22]:
def group_func(data, group, ind, debt='debt'):
    data_pivot = data.pivot_table(index=group, values=debt, aggfunc=['count', 'sum', 'mean']) 
    data_pivot.columns = ['Всего кредитополучателей', 'Всего должников', 'Доля должников']
    data_pivot.index.names = [ind]
    data_pivot = data_pivot.sort_values(by='Доля должников',ascending=False)
    print(data_pivot)

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Чтобы узнать ответ, сделаем группировку данных по количеству детей (столбец children).

In [23]:
group_func(data, 'children', 'Кол-во детей')

              Всего кредитополучателей  Всего должников  Доля должников
Кол-во детей                                                           
4                                   41                4        0.097561
2                                 2052              194        0.094542
1                                 4808              444        0.092346
3                                  330               27        0.081818
0                                14091             1063        0.075438
5                                    9                0        0.000000


**Вывод:** 

Несмотря на то, что люди без детей берут больше кредитов, доля задолженностей у них меньше, чем у родителей 1-4 детей. С 4-мя детьми доля задолженностей больше всех. Тенденция роста долей у людей до 2х детей (вкл) показывает зависимость данных (чем больше делей, тем больше доля должников). Однако кредитов у  многодетных, особенно с 4 и 5, меньше, что может указать на то что мало данных. Возможно, люди с детьми (по данным) берут меньше кредитов по каким-то обстоятельствам, например, выплачивать тяжелее.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Воспользуемся функцией для создания таблицы по семейному положению (family_status).

In [24]:
group_func(data, 'family_status', 'Семейное положение')

                       Всего кредитополучателей  Всего должников  \
Семейное положение                                                 
Не женат / не замужем                      2796              273   
гражданский брак                           4134              385   
женат / замужем                           12261              927   
в разводе                                  1189               84   
вдовец / вдова                              951               63   

                       Доля должников  
Семейное положение                     
Не женат / не замужем        0.097639  
гражданский брак             0.093130  
женат / замужем              0.075606  
в разводе                    0.070648  
вдовец / вдова               0.066246  


**Вывод:** 

Большую часть кредитов берут люди, которые «женаты/замужем». Скорее всего, условия достаточно приятны для семейных пар (которые официально в браке), нежели для одиночек («не женат/не замужем» 2796, «в разводе» 1189, «вдовец/вдова» 951), а может, малая выборка по другим категориям семейного положения. О лучших условиях для женатых людей еще показывает значительная доля задолженностей по семейному положению (на первом месте «не женат / не замужем»). 

Незначительная зависимость между семейным положением и возвратом кредита в срок имеется. Люди, которые не были в официальном браке, примерно на 2% чаще не возвращают по кредиту (9,76% неженатые люди и 9,31% в гражданском браке, остальные менее 7,5%).

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Воспользуемся функцией для создания таблицы по уровню дохода (total_income_category):
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

In [25]:
group_func(data, 'total_income_category', 'Группа дохода')

               Всего кредитополучателей  Всего должников  Доля должников
Группа дохода                                                           
E                                    22                2        0.090909
C                                 15921             1353        0.084982
A                                    25                2        0.080000
B                                  5014              354        0.070602
D                                   349               21        0.060172


**Вывод:** 

У людей с самым высоким и самым низким уровнем дохода меньше всего кредитов, что достаточно понятно (люди с низким доходом не могут себе позволить кредит, а людям с высоким он, скорее всего, не очень нужен). Люди категории «С» и «B», — это топ-2 по кол-ву взятых кредитов, данных по остальным категориям мало. 

Категория «С» самая многочисленная, но у нее выше доля должников, чем у категории «B» (8,5% и 7,06% соответственно). Это может быть связано с тем, что у большего числа населения (зарплаты 50001–200000 — это большая масса работающего населения) менее стабильное финансовое положение. Тогда как категория «B» показывает меньший риск по задолженности по кредиту, возможно, из-за большей финансовой стабильности.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Воспользуемся функцией для создания таблицы по целям кредита (purpose_category).

In [26]:
group_func(data, 'purpose_category', 'Цель кредита')

                          Всего кредитополучателей  Всего должников  \
Цель кредита                                                          
операции с автомобилем                        4279              400   
получение образования                         3988              369   
проведение свадьбы                            2313              183   
операции с недвижимостью                     10751              780   

                          Доля должников  
Цель кредита                              
операции с автомобилем          0.093480  
получение образования           0.092528  
проведение свадьбы              0.079118  
операции с недвижимостью        0.072551  


**Вывод:** 

Операции с недвижимостью в топе по кол-ву взятых кредитов, а также у них самая малая доля задолженностей - это логично, недвижимость дорогая, выплачивать ипотеку долго и части ежемесячного платежа меньше, чем у других типов кредита. 

Операции с автомобилем и получение образования в одинаковом среднем сегменте по кол-ву взятых кредитов и по кол-ву задолженностей. Также их доля задолженностей выше, поэтому можно смело предположить, что гипотеза «Люди, берущие кредит для операций с автомобилем и получения образования, чаще сталкиваются с задолженностями по сравнению с людьми, берущими кредит на другие цели (свадьба или недвижимость)» верна.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* С учетом того, что пропуски были в общем трудовом стаже и ежемесячном доходе, смею предположить, что этих данных может не быть у клиента, который не работает или работает неофициально (например, пенсионер), либо эти данные просто не указали сами клиенты по своим причинам. Если эта информация заполняется из карты клиентов, а не клиентами, то, вероятно, клиенты с отсутствующими данными не являются зарплатными клиентами или произошел сбой сбора данных. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что значение средних может быть искажено выбросами данных (те которые очень большие или наоборот очень маленькие по сравнению с основной массой), а медиана это ровно половина элементов.

### Шаг 4: общий вывод.

**Предобработка**

Ознакомившись с данными было выявлено некорректное заполнение данных в столбце days_employed (общий трудовой стаж в днях), а именно отрицательные значения. Так как стаж не может быть отрицательным, то поправила данные с помощью abs(). Далее нашла 2174 пустых значений в столбцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход). Так как данные количественные и могут содержать выбросы, то заполнила эти пробелы медианным значением по указанному типу занятости. Далее были найдены и убраны некорректные строки по анамальному значению кол-ва детей, а именно отрицательные и 20. 

**Разбивка данных на категории для удобства восприятия**


1. Данные в столбце ежемесячного дохода были сгруппированы по категориям дохода в новый столбец total_income_category, для этого я перевела данные столбца в целочисленный тип, чтобы можно было совершать действия как с числами.

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

2. Также был разбит на категории столбец purpose (цель получения кредита) на 4 группы:


- операции с автомобилем
- операции с недвижимостью
- проведение свадьбы
- получение образования

**Выявлено**

1. Зависимость задолженности от количества детей у клиента: чем больше детей, тем выше доля должников (по данным с 0 до 2 включительно детей).
2. Люди, которые были в официальном браке, надежнее, их доля задолженности составляет 7,56%. У людей не женатых и в гражданском браке доля выше — 9,76% и 9,31%.
3. Клиенты категорий дохода 'C' и 'B берут большую часть кредитов, однако категория 'B' более надежная. Доли по группам 'C': 8,5%, 'B': 7,06%.
4. Люди, берущие кредит для операций с автомобилем и получения образования, чаще сталкиваются с задолженностями по сравнению с людьми, берущими кредит на другие цели (свадьба или недвижимость).

**Вывод**

Самый надежный кандидат на кредит по данным — человек состоявший в официальном браке, из категории дохода 'B' и желательно без детей или с 1 ребенком.

Самый рискованный заемщик по данным — этот человек не женат/ не замужем или в гражданском браке с детьми от 2-х и более, из категории дохода 'C' и ниже.

Также кредиты для проведения свадьбы или операций с недвижимостью стабильнее, так как доля задолженности в них ниже. 